turbo/engineapi/fork_validator.go

```diff
@@ -0,0 +1,213 @@
+   You may obtain a copy of the License at
+       http://www.apache.org/licenses/LICENSE-2.0
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package engineapi
+
+import (
+	"fmt"
+
+	"github.com/ledgerwatch/erigon-lib/gointerfaces/remote"
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon-lib/kv/memdb"
+	"github.com/ledgerwatch/erigon/common"
+	"github.com/ledgerwatch/erigon/core/rawdb"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/log/v3"
+)
+
+// the maximum point from the current head, past which side forks are not validated anymore.
+const maxForkDepth = 32 // 32 slots is the duration of an epoch thus there cannot be side forks in PoS deeper than 32 blocks from head.
+
+type validatePayloadFunc func(kv.RwTx, *types.Header, *types.RawBody, uint64, []*types.Header, []*types.RawBody) error
+
+// Fork segment is a side fork segment and repressent a full side fork block.
+type forkSegment struct {
+	header *types.Header
+	body   *types.RawBody
+}
+
+type ForkValidator struct {
+	// Hash => side fork block, any block saved into this map is considered valid.
+	// blocks saved are required to have at most distance maxForkDepth from the head.
+	// if we miss a segment, we only accept the block and give up on full validation.
+	sideForksBlock map[common.Hash]forkSegment
+	// current memory batch containing chain head that extend canonical fork.
+	extendingFork *memdb.MemoryMutation
+	// hash of chain head that extend canonical fork.
+	extendingForkHeadHash common.Hash
+	// this is the function we use to perform payload validation.
+	validatePayload validatePayloadFunc
+}
+
+// abs64 is a utility method that given an int64, it returns its absolute value in uint64.
+func abs64(n int64) uint64 {
+	if n < 0 {
+		return uint64(-n)
+	}
+	return uint64(n)
+}
+
+func NewForkValidatorMock() *ForkValidator {
+	return &ForkValidator{
+		sideForksBlock: make(map[common.Hash]forkSegment),
+	}
+}
+
+func NewForkValidator(validatePayload validatePayloadFunc) *ForkValidator {
+	return &ForkValidator{
+		sideForksBlock:  make(map[common.Hash]forkSegment),
+		validatePayload: validatePayload,
+	}
+}
+
+// ExtendingForkHeadHash return the fork head hash of the fork that extends the canonical chain.
+func (fv *ForkValidator) ExtendingForkHeadHash() common.Hash {
+	return fv.extendingForkHeadHash
+}
+
+// FlushExtendingFork flush the current extending fork if fcu chooses its head hash as the its forkchoice.
+func (fv *ForkValidator) FlushExtendingFork(tx kv.RwTx) error {
+	// Flush changes to db.
+	if err := fv.extendingFork.Flush(tx); err != nil {
+		return err
+	}
+	// Clean extending fork data
+	if fv.extendingFork != nil {
```

Comment:
`Flush()` followed by `Rollback()` looks funny. To my mind `Flush` sounds like transaction commit, thus it shouldn't be followed by rollback.

Also, there's no point in `fv.extendingFork != nil` check after calling `fv.extendingFork.Flush(tx)`.

Reply:
but yes the if was a mistake